crucible-llvm/src/Lang/Crucible/LLVM/MemModel.hs

```diff
@@ -1501,3 +1504,55 @@ allocGlobal sym mem (g, aliases, sz, alignment) = do
   -- TODO: Aliases are not propagated to doMalloc for error messages
   (ptr, mem') <- doMalloc sym G.GlobalAlloc mut sym_str mem sz' alignment
   return (registerGlobal mem' (symbol:aliases) ptr)
+
+-- | Look up a pointer in the 'memImplGlobalMap' to see if it's a global.
+--
+-- This is best-effort and will only work if the pointer is fully concrete
+-- and matches the address of the global on the nose. It is used in SAWscript
+-- for friendly error messages.
+isGlobalPointer :: forall sym w. (IsSymInterface sym, 1 <= w)
+                => sym
+                -> GlobalMap sym
+                -> LLVMPtr sym w
+                -> IO (Maybe L.Symbol)
+isGlobalPointer sym globalMap needle =
+  let fun :: L.Symbol -> SomePointer sym -> IO (Maybe L.Symbol)
+      fun symb (SomePointer ptr)
+        | Just Refl <- testEquality ptr needle -- do they have the same width?
+        = ptrEq sym (ptrWidth ptr) ptr needle <&> \equalityPredicate ->
+            case asConstantPred equalityPredicate of
+              Just True  -> Just symb
+              Just False -> Nothing
+              Nothing    -> Nothing
+      fun _ _ = pure Nothing
+  in listToMaybe . fmap fst . Map.toList <$>
+       Map.traverseMaybeWithKey fun globalMap
+
+-- | For when you don't know @1 <= w@
+isGlobalPointer' :: forall sym w. (IsSymInterface sym)
+                 => sym
+                 -> GlobalMap sym
+                 -> LLVMPtr sym w
+                 -> IO (Maybe L.Symbol)
+isGlobalPointer' sym globalMap needle =
+  case testLeq (knownNat :: NatRepr 1) (ptrWidth needle) of
+    Nothing       -> pure Nothing
+    Just LeqProof -> isGlobalPointer sym globalMap needle
+
+-- | Pretty-print an 'LLVMVal', but replace pointers to globals with the name of
+--   the global when possible. Probably pretty slow on big structures.
+--
+-- TODO(lb): This is in this module instead of "MemModel.Value" because
```

Comment:
I could use some help here if someone wouldn't mind taking a look. Thanks in advance!